File: aoj/0615-2.rb
$memo = {}
$n = gets.to_i
$a = (1..$n).map{gets.to_i}

def inc(x);(x+1)%$n;end
def dec(x);(x+$n-1)%$n;end

def dfs(l,r)
  $memo[[l,r]] ||= if l==r
    $a[l]
  elsif inc(l)==r
    [$a[l],$a[r]].max
  else
    [[$a[l],inc(l),r],[$a[r],l,dec(r)]].map{|x,nextl,nextr|
      if $a[nextl] > $a[nextr]
        nextl = inc(nextl)
      else
        nextr = dec(nextr)
      end
      x + dfs(nextl,nextr)
    }.max
  end
end

p $n.times.map{|i|dfs(inc(i),i)}.max

File: aoj/0615-2.py
n = int(input())
a = [int(input()) for _ in range(n)]
dp = [[0]*n for _ in range(n)]

for i in range(n):
    if i%2 == n%2:
        for l,r in [(t,(t+i)%n) for t in range(n)]:
            dp[l][r] = dp[(l+1)%n][r] if a[l]>a[r] else dp[l][(r+n-1)%n]
    else:
        for l,r in [(t,(t+i)%n) for t in range(n)]:
            dp[l][r] = max(a[l] + dp[(l+1)%n][r],a[r] + dp[l][(r+n-1)%n])

print(max(dp[(i+1)%n][i] for i in range(n)))
